File: backend/services/webhook-dispatcher/signing/signature.go
// Package signing implements only the F07 customer-webhook v1 proof contract.
// It does not verify Meta/Telegram callbacks, authorize a tenant, send HTTP,
// claim a replay receipt, or acknowledge durable acceptance. Those owners remain
// separate. Keys must come from the authenticated endpoint/environment context.
package signing

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"net/http"
	"regexp"
	"strconv"
	"strings"
)

// Stable errors contain no headers, keys or payload fragments. Hosts map proof
// rejection to a neutral response; a configuration error is an operator fault.
var (
	ErrConfig = errors.New("WEBHOOK_SIGNING_CONFIG_INVALID")
	ErrProof  = errors.New("WEBHOOK_PROOF_INVALID")
)

const (
	EventHeader     = "X-Platform-Event-Id"
	DeliveryHeader  = "X-Platform-Delivery-Id"
	TimestampHeader = "X-Platform-Timestamp"
	KeyHeader       = "X-Platform-Key-Id"
	SignatureHeader = "X-Platform-Signature"
	maxUnix         = int64(9999999999)
)

var (
	// A delivery ID cannot contain the '.' framing delimiter. UUIDs and the
	// baseline's opaque example IDs are accepted without numeric coercion.
	identity  = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9_:-]{0,127}$`)
	keyID     = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9_.:-]{0,127}$`)
	timestamp = regexp.MustCompile(`^(0|[1-9][0-9]{0,9})$`)
	signature = regexp.MustCompile(`^v1=[0-9a-f]{64}$`)
)

// Policy is supplied by endpoint configuration, never webhook headers. No
// default replay window is silently chosen. Limits bound CPU/memory only and
// are not evidence that a production endpoint or retention policy is approved.
type Policy struct {
	MaxAgeSeconds        int64
	MaxFutureSkewSeconds int64
	MaxBodyBytes         int
}

// KeyVersion describes a route-owned current or previous key. Time intervals
// are [NotBefore, SignUntil) for attempts and [NotBefore, VerifyUntil) for
// verification. Rotation stops old signing before bounded verification overlap
// ends. Revocation takes precedence over both intervals. Do not log this struct.
// Callers must not mutate Secret concurrently with Sign or Verify.
type KeyVersion struct {
	ID          string
	Secret      []byte
	NotBefore   int64
	SignUntil   int64
	VerifyUntil int64
	Revoked     bool
}

// Proof reports authentication of exact bytes and delivery identity only. The
// unsigned EventHeader is intentionally absent. Consumers must validate an event
// ID in the authenticated body and compare that to the header before using it.
// Retain DeliveryID + BodySHA256 in an endpoint-scoped durable receipt to detect
// same-delivery/different-body conflicts; the timestamp is NOT a deduplication key.
type Proof struct {
	DeliveryID  string `json:"delivery_id"`
	KeyID       string `json:"key_id"`
	AttemptedAt int64  `json:"attempted_at"`
	BodySHA256  string `json:"body_sha256"`
}

func validPolicy(p Policy, now int64) bool {
	return now >= 0 && now <= maxUnix && p.MaxAgeSeconds >= 1 &&
		p.MaxAgeSeconds <= 86400 && p.MaxFutureSkewSeconds >= 0 &&
		p.MaxFutureSkewSeconds <= p.MaxAgeSeconds && p.MaxBodyBytes >= 1 && p.MaxBodyBytes <= 1048576
}

func validKey(k KeyVersion) bool {
	return keyID.MatchString(k.ID) && len(k.Secret) >= 32 && len(k.Secret) <= 4096 &&
		k.NotBefore >= 0 && k.NotBefore < k.SignUntil && k.SignUntil <= k.VerifyUntil && k.VerifyUntil <= maxUnix
}

func mac(body []byte, deliveryID, at string, secret []byte) []byte {
	h := hmac.New(sha256.New, secret)
	_, _ = h.Write([]byte(at + "." + deliveryID + "."))
	_, _ = h.Write(body)
	return h.Sum(nil)
}

// Sign authenticates timestamp + '.' + deliveryID + '.' + exact raw body.
// body: serialized once by the caller; send these SAME bytes, not a JSON rebuild.
// eventID/deliveryID: stable IDs from an authorized durable delivery record, not
// caller-supplied tenant authority. A retry keeps deliveryID but gets a fresh now.
// key/policy/now: route-owned key, explicit limits and trusted Unix-second clock.
// Returns the five F07 headers or a stable error. No I/O, ledger write, logging,
// acknowledgement, key generation or runtime registration occurs here.
func Sign(body []byte, eventID, deliveryID string, key KeyVersion, now int64, policy Policy) (http.Header, error) {
	if !validPolicy(policy, now) || !validKey(key) {
		return nil, ErrConfig
	}
	if len(body) > policy.MaxBodyBytes || !identity.MatchString(eventID) || !identity.MatchString(deliveryID) ||
		key.Revoked || now < key.NotBefore || now >= key.SignUntil {
		return nil, ErrProof
	}
	at := strconv.FormatInt(now, 10)
	headers := make(http.Header, 5)
	headers.Set(EventHeader, eventID)
	headers.Set(DeliveryHeader, deliveryID)
	headers.Set(TimestampHeader, at)
	headers.Set(KeyHeader, key.ID)
	headers.Set(SignatureHeader, "v1="+hex.EncodeToString(mac(body, deliveryID, at, key.Secret)))
	return headers, nil
}

// one rejects duplicates, including differently cased map keys. Calling Header.Get
// alone would hide repeated security headers. HTTP adapters must preserve duplicates.
func one(headers http.Header, name string) (string, bool) {
	value, count := "", 0
	for k, values := range headers {
		if strings.EqualFold(k, name) {
			count += len(values)
			if len(values) > 0 {
				value = values[0]
			}
		}
	}
	return value, count == 1
}

// Verify checks exact proof framing, key validity/rotation/revocation, a bounded
// trusted-clock window and constant-time MAC equality. keys is the endpoint's
// already authorized ring of one or two keys; NEVER resolve a URL/key from headers.
// body/headers are bounded untrusted request bytes; now and policy are host values.
// Returns authenticated delivery metadata, NOT replay protection, a durable ACK,
// a validated event schema, or business completion. All proof failures are neutral.
// The caller must not send 2xx until its separate durable acceptance boundary passes.
func Verify(body []byte, headers http.Header, keys []KeyVersion, now int64, policy Policy) (Proof, error) {
	empty := Proof{}
	if !validPolicy(policy, now) || len(keys) < 1 || len(keys) > 2 {
		return empty, ErrConfig
	}
	for i, k := range keys {
		if !validKey(k) {
			return empty, ErrConfig
		}
		for j := 0; j < i; j++ {
			if k.ID == keys[j].ID || hmac.Equal(k.Secret, keys[j].Secret) {
				return empty, ErrConfig
			}
		}
	}
	if len(body) > policy.MaxBodyBytes || len(headers) > 128 {
		return empty, ErrProof
	}
	event, a := one(headers, EventHeader)
	delivery, b := one(headers, DeliveryHeader)
	at, c := one(headers, TimestampHeader)
	kid, d := one(headers, KeyHeader)
	sig, e := one(headers, SignatureHeader)
	if !a || !b || !c || !d || !e || !identity.MatchString(event) || !identity.MatchString(delivery) ||
		!timestamp.MatchString(at) || !keyID.MatchString(kid) || !signature.MatchString(sig) {
		return empty, ErrProof
	}
	t, err := strconv.ParseInt(at, 10, 64)
	if err != nil || t > maxUnix || (t <= now && now-t > policy.MaxAgeSeconds) ||
		(t > now && t-now > policy.MaxFutureSkewSeconds) {
		return empty, ErrProof
	}
	var selected *KeyVersion
	for i := range keys {
		if keys[i].ID == kid {
			selected = &keys[i]
		}
	}
	if selected == nil || selected.Revoked || t < selected.NotBefore || t >= selected.SignUntil ||
		now < selected.NotBefore || now >= selected.VerifyUntil {
		return empty, ErrProof
	}
	supplied, _ := hex.DecodeString(sig[3:]) // shape already checked; fixed 32-byte input
	if !hmac.Equal(supplied, mac(body, delivery, at, selected.Secret)) {
		return empty, ErrProof
	}
	digest := sha256.Sum256(body)
	return Proof{DeliveryID: delivery, KeyID: kid, AttemptedAt: t, BodySHA256: hex.EncodeToString(digest[:])}, nil
}

File: backend/sdk/python/greeto_webhooks/signature.py
"""F07 customer-webhook v1 helpers; NOT a Meta/Telegram verifier.

Only exact-byte authentication is implemented. No I/O, key lookup, tenant grant,
replay receipt, HTTP acknowledgement or business effect is performed. Endpoint
keys and clocks must be supplied by the authorized server, never the request.
"""
from __future__ import annotations
from dataclasses import dataclass
import hashlib
import hmac
import re
from types import MappingProxyType
from typing import Mapping, Sequence

HEADERS = MappingProxyType({'event': 'X-Platform-Event-Id', 'delivery': 'X-Platform-Delivery-Id',
    'timestamp': 'X-Platform-Timestamp', 'key': 'X-Platform-Key-Id', 'signature': 'X-Platform-Signature'})
_ID = re.compile(r'[A-Za-z0-9][A-Za-z0-9_:-]{0,127}')
_KEY = re.compile(r'[A-Za-z0-9][A-Za-z0-9_.:-]{0,127}')
_TIME = re.compile(r'(0|[1-9][0-9]{0,9})')
_SIGNATURE = re.compile(r'v1=[0-9a-f]{64}')
_MAX_UNIX = 9999999999


@dataclass(frozen=True)
class Policy:
    """Explicit seconds/byte limits; no replay window is silently selected."""
    max_age_seconds: int
    max_future_skew_seconds: int
    max_body_bytes: int


@dataclass(frozen=True)
class KeyVersion:
    """Endpoint/environment-owned random bytes, never a password or request key.

    Signing interval is [not_before, sign_until); verification overlap ends at
    verify_until exclusively. Revoked keys cannot sign or verify. Do not log keys.
    """
    id: str
    secret: bytes
    not_before: int
    sign_until: int
    verify_until: int
    revoked: bool = False


@dataclass(frozen=True)
class Proof:
    """Authenticated delivery/body only. The event-ID header is NOT proof data.

    Persist delivery_id + body_sha256 under authenticated endpoint scope for
    conflict/replay checks. A valid signature does not supply a durable receipt.
    """
    delivery_id: str
    key_id: str
    attempted_at: int
    body_sha256: str


class WebhookProofError(ValueError):
    """Payload-free stable code for neutral HTTP mapping by the owning handler."""
    def __init__(self, config: bool = False) -> None:
        self.code = 'WEBHOOK_SIGNING_CONFIG_INVALID' if config else 'WEBHOOK_PROOF_INVALID'
        super().__init__(self.code)


def _integer(n: object, lo: int, hi: int) -> bool:
    return type(n) is int and lo <= n <= hi


def _matches(pattern: re.Pattern[str], value: object) -> bool:
    return isinstance(value, str) and pattern.fullmatch(value) is not None


def _policy(p: Policy, now: int) -> None:
    if not isinstance(p, Policy) or not _integer(now, 0, _MAX_UNIX) or not (
        _integer(p.max_age_seconds, 1, 86400) and
        _integer(p.max_future_skew_seconds, 0, p.max_age_seconds) and
        _integer(p.max_body_bytes, 1, 1048576)):
        raise WebhookProofError(True)


def _key(k: KeyVersion) -> None:
    if not isinstance(k, KeyVersion) or not _matches(_KEY, k.id) or not (
        isinstance(k.secret, bytes) and 32 <= len(k.secret) <= 4096 and
        _integer(k.not_before, 0, _MAX_UNIX) and _integer(k.sign_until, k.not_before + 1, _MAX_UNIX) and
        _integer(k.verify_until, k.sign_until, _MAX_UNIX) and type(k.revoked) is bool):
        raise WebhookProofError(True)


def _mac(body: bytes, delivery: str, at: str, secret: bytes) -> bytes:
    digest = hmac.new(secret, f'{at}.{delivery}.'.encode('ascii'), hashlib.sha256)
    digest.update(body)
    return digest.digest()


def sign_webhook(raw_body: bytes, event_id: str, delivery_id: str, key: KeyVersion,
                 now: int, policy: Policy) -> Mapping[str, str]:
    """Return F07 headers for authorized, already-durable delivery state.

    raw_body: immutable serialized bytes; send precisely these bytes. event_id
    and delivery_id: stable opaque IDs, with no '.' framing delimiter. key/now/
    policy: trusted endpoint key, Unix-second clock and explicit bounds. Retry
    keeps delivery_id but refreshes now. No send, receipt write or ACK occurs.
    Raises WebhookProofError for config, bounds, key intervals or revocation.
    """
    _policy(policy, now)
    _key(key)
    if not isinstance(raw_body, bytes) or len(raw_body) > policy.max_body_bytes or not (
        _matches(_ID, event_id) and _matches(_ID, delivery_id)) or key.revoked or not (
        key.not_before <= now < key.sign_until):
        raise WebhookProofError()
    at = str(now)
    return MappingProxyType({HEADERS['event']: event_id, HEADERS['delivery']: delivery_id,
        HEADERS['timestamp']: at, HEADERS['key']: key.id,
        HEADERS['signature']: 'v1=' + _mac(raw_body, delivery_id, at, key.secret).hex()})


def verify_webhook(raw_body: bytes, raw_headers: Sequence[tuple[str, str]],
                   endpoint_keys: Sequence[KeyVersion], now: int, policy: Policy) -> Proof:
    """Verify raw bytes, clock window, one/two route-owned keys and exact proof.

    raw_headers must preserve pairs/duplicates; a collapsed dictionary is rejected.
    endpoint_keys come from authenticated endpoint/environment config, not request
    IDs/URLs. Returns Proof only; no event schema, grant, replay ledger or ACK is
    certified. The receiver validates the signed body/event ID and durably records
    its scoped receipt/effect before 2xx. Raw timestamps are not deduplication keys.
    Raises neutral WebhookProofError; never echoes input/secret material.
    """
    _policy(policy, now)
    if not isinstance(endpoint_keys, (list, tuple)) or not 1 <= len(endpoint_keys) <= 2:
        raise WebhookProofError(True)
    for k in endpoint_keys:
        _key(k)
    if len(endpoint_keys) == 2 and (endpoint_keys[0].id == endpoint_keys[1].id or
        hmac.compare_digest(endpoint_keys[0].secret, endpoint_keys[1].secret)):
        raise WebhookProofError(True)
    if not isinstance(raw_body, bytes) or len(raw_body) > policy.max_body_bytes or not (
        isinstance(raw_headers, (list, tuple)) and len(raw_headers) <= 128):
        raise WebhookProofError()
    for pair in raw_headers:
        if not isinstance(pair, (tuple, list)) or len(pair) != 2 or any(
            not isinstance(v, str) or len(v) > 8192 for v in pair):
            raise WebhookProofError()
    def one(name: str) -> str:
        values = [v for k, v in raw_headers if k.lower() == name.lower()]
        if len(values) != 1:
            raise WebhookProofError()
        return values[0]
    event, delivery, at, kid, sig = (one(HEADERS[k]) for k in ('event', 'delivery', 'timestamp', 'key', 'signature'))
    if not (_matches(_ID, event) and _matches(_ID, delivery) and _matches(_TIME, at) and
            _matches(_KEY, kid) and _matches(_SIGNATURE, sig)):
        raise WebhookProofError()
    attempted_at = int(at)
    if (attempted_at <= now and now - attempted_at > policy.max_age_seconds) or (
        attempted_at > now and attempted_at - now > policy.max_future_skew_seconds):
        raise WebhookProofError()
    selected = next((k for k in endpoint_keys if k.id == kid), None)
    if selected is None or selected.revoked or not (selected.not_before <= attempted_at < selected.sign_until and
                                                   selected.not_before <= now < selected.verify_until):
        raise WebhookProofError()
    if not hmac.compare_digest(bytes.fromhex(sig[3:]), _mac(raw_body, delivery, at, selected.secret)):
        raise WebhookProofError()
    return Proof(delivery, kid, attempted_at, hashlib.sha256(raw_body).hexdigest())
